src/transaction/transaction.ts

```diff
@@ -1,45 +1,128 @@
+const _rawTxToTxSigned = (
+  params: SigningParameters,
   byronWitnesses: TxWitnessByron[],
   shelleyWitnesses: TxWitnessShelley[],
 ): TxCborHex => {
-  const { body, scriptWitnesses, auxiliaryData } = rawTx
-  const nativeScriptWitnessList = scriptWitnesses as any[] | undefined
-  const witnessSet = new Map()
+  const {
+    body, scriptWitnesses, datumWitnesses, redeemerWitnesses, scriptValidity, auxiliaryData,
+  } = params.rawTx!
+
+  // cardano-cli deduplicates the script witnesses before adding them to the signed transaction
+  const scriptWitnessList = uniqWith((scriptWitnesses || []) as any[], isEqual)
+  const datumWitnessList = (datumWitnesses || []) as any[]
+  const redeemerWitnessList = (redeemerWitnesses || []) as any[]
+
+  let nativeScriptWitnessList: unknown[] = []
+  let plutusScriptWitnessList: unknown[] = []
+  if (params.era === CardanoEra.ALONZO) {
+    // In alonzo era txs, each native script is packed in a list with 0 literal, each plutus script
+    // is packed in a list with 1 literal and they are all stored in scriptWitnessList.
+    // We need to partition them back and unpack the actual scripts.
+    nativeScriptWitnessList = scriptWitnessList.filter((s) => s[0] === 0).map((s) => s[1])
+    plutusScriptWitnessList = scriptWitnessList.filter((s) => s[0] === 1).map((s) => s[1])
+  } else {
+    // In older txs, scriptWitnessList is just a list of native scripts.
+    nativeScriptWitnessList = scriptWitnessList
+  }
+
+  const witnessSet: WitnessSet = new Map()
+
   if (shelleyWitnesses.length > 0) {
     witnessSet.set(TxWitnessKeys.SHELLEY, shelleyWitnesses)
   }
-  if (nativeScriptWitnessList?.length) {
-    // cardano-cli deduplicates the script witnesses before adding them to the signed transaction
-    witnessSet.set(TxWitnessKeys.NATIVE_SCRIPTS, uniqWith(nativeScriptWitnessList, isEqual))
+
+  if (nativeScriptWitnessList.length > 0) {
+    witnessSet.set(TxWitnessKeys.NATIVE_SCRIPTS, nativeScriptWitnessList)
   }
+
   if (byronWitnesses.length > 0) {
     witnessSet.set(TxWitnessKeys.BYRON, byronWitnesses)
   }
-  return encodeTx({ body, witnessSet, auxiliaryData }).toString('hex') as TxCborHex
+
+  if (plutusScriptWitnessList.length > 0) {
+    witnessSet.set(TxWitnessKeys.PLUTUS_SCRIPTS, plutusScriptWitnessList)
+  }
+
+  if (datumWitnessList.length > 0) {
+    witnessSet.set(TxWitnessKeys.PLUTUS_DATA, datumWitnessList)
+  }
+
+  if (redeemerWitnessList.length > 0) {
+    witnessSet.set(TxWitnessKeys.REDEEMERS, redeemerWitnessList)
+  }
+
+  const signedTx = {
+    body, witnessSet, scriptValidity, auxiliaryData,
+  }
+  return InteropLib.encodeTx(signedTx).toString('hex') as TxCborHex
+}
+
+const TxSigned = (
+  params: SigningParameters,
+  byronWitnesses: TxWitnessByron[],
+  shelleyWitnesses: TxWitnessShelley[],
+): TxCborHex => {
+  if (params.rawTx) {
+    return _rawTxToTxSigned(params, byronWitnesses, shelleyWitnesses)
+  }
+
+  const witnessSet = _parseWitnessSet(params.tx!.witnessSet)
```

Comment:
I'd welcome a comment explaining we only need to add witnesses provided by the HW wallet and everything else (e.g. scripts) stays the same.

Reply:
Added